### PR TITLE
Icc stuff

### PR DESCRIPTION
**Pull request by fweik**
Description of changes:
 - Made icc initialization faster
 - Fixed ghost charge synchronization issue
 - Cleanup
 - Allow elc and icc to be used together

### CODE REVIEW

File: src/core/electrostatics_magnetostatics/icc.cpp [hunk 1]
```diff
@@ -126,194 +121,118 @@ inline void add_non_bonded_pair_force_iccp3m(Particle *p1, Particle *p2,
...
+    return 0;
 
   iccp3m_sanity_check();
 
   if ((iccp3m_cfg.eout <= 0)) {
-    ostringstream msg;
-    msg << "ICCP3M: nonpositive dielectric constant is not allowed. Put a "
+    runtimeErrorMsg()
+        << "ICCP3M: nonpositive dielectric constant is not allowed. Put a "
            "decent exception here\n";
```
Comment (KonradBreitsprecher) [inline src/core/electrostatics_magnetostatics/icc.cpp]: Decent enough I guess

**Comment by fweik** [comment]
Gitlab passed (https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/pipelines/3518), not sure why it is now showing.

**Comment by mkuron** [comment]
@fweik, we had a Gitlab outage this morning, so half an hour's worth of pull requests was not mirrored. You have to add another commit if you want to rebuild this pull request. The link you posted is unrelated and for #1924.

File: src/core/electrostatics_magnetostatics/icc.cpp [hunk 2]
```diff
@@ -126,194 +121,118 @@ inline void add_non_bonded_pair_force_iccp3m(Particle *p1, Particle *p2,
...
-      } /* cell particles */
-    }   /* local cells */
+      }
+    } /* cell particles */
+    /* Update charges on ghosts. */
+    ghost_communicator(&cell_structure.exchange_ghosts_comm);
+
     iccp3m_cfg.citeration++;
+
     MPI_Allreduce(&diff, &globalmax, 1, MPI_DOUBLE, MPI_MAX, comm_cart);
```
Comment (KonradBreitsprecher) [inline src/core/electrostatics_magnetostatics/icc.cpp:219]: ICC_DEBUG is unused, I suggest to add sth like this here:
```
#ifdef ICC_DEBUG
    fprintf(stderr, "Iteration %d globalmax %f\n", iccp3m_cfg.citeration, globalmax);
#endif
```

File: src/core/electrostatics_magnetostatics/icc.cpp [hunk 3]
```diff
@@ -126,194 +121,118 @@ inline void add_non_bonded_pair_force_iccp3m(Particle *p1, Particle *p2,
...
+            (1. - iccp3m_cfg.relax) * hold + (iccp3m_cfg.relax) * (f1 + f2);
+
+        /* Take the largest error to check for convergence */
+        auto const relative_difference =
+            std::abs(1 * (hnew - hold) / (hmax + std::abs(hnew + hold)));
+
+        diff = std::max(diff, relative_difference);
+
+        p.p.q = hnew * iccp3m_cfg.areas[id];
+
```
Comment (KonradBreitsprecher) [inline src/core/electrostatics_magnetostatics/icc.cpp:197]: ICC_DEBUG is unused, I suggest to add sth like this here:

```
#ifdef ICC_DEBUG
   fprintf(stderr, "Particle %d q_old %10.5f q_new %10.5f F_n %10.5f E %10.5f %10.5f %10.5f\n", p.p.identity, hold * iccp3m_cfg.areas[id], p.p.q, f1, ex,ey,ez);
#endif
```

**Review by KonradBreitsprecher — CHANGES_REQUESTED** [review]
Much cleaner. About deactivating ICC, I guess it's ok to leave the ICC charges on their last value, resetting to zero can be easily done on python level.
I'd like to have the ICC_DEBUG stuff in, otherwise it's fine. ICC+ELC needs testing however, see issue #2278.

File: src/core/electrostatics_magnetostatics/icc.cpp [hunk 4]
```diff
@@ -126,194 +121,118 @@ inline void add_non_bonded_pair_force_iccp3m(Particle *p1, Particle *p2,
...
+      if (p.p.identity < iccp3m_cfg.n_ic + iccp3m_cfg.first_id &&
+          p.p.identity >= iccp3m_cfg.first_id) {
+        auto const id = p.p.identity - iccp3m_cfg.first_id;
+        /* the dielectric-related prefactor: */
+        auto const del_eps = (iccp3m_cfg.ein[id] - iccp3m_cfg.eout) /
+                             (iccp3m_cfg.ein[id] + iccp3m_cfg.eout);
+        /* calculate the electric field at the certain position */
+        auto const ex = p.f.f[0] / p.p.q + iccp3m_cfg.extx;
+        auto const ey = p.f.f[1] / p.p.q + iccp3m_cfg.exty;
+        auto const ez = p.f.f[2] / p.p.q + iccp3m_cfg.exty;
```
Comment (KonradBreitsprecher) [inline src/core/electrostatics_magnetostatics/icc.cpp]: Should be ext**z** here.

**Comment by fweik** [comment]
Good catch.

**Comment by fweik** [comment]
Do you see any benefit for supporting MMM2D with ICC when you could use ELC instead? It seems to me that ELC supports more boundary conditions and is faster...

**Comment by KonradBreitsprecher** [comment]
Not really, but if we figured out the validity of 2D methods + ICC, why would you take it out?

**Comment by fweik** [comment]
Wouldn't it be better to collect proper statistics (maybe on demand) than the debug statements? How do you use this?

**Comment by KonradBreitsprecher** [comment]
I collect the number of iterations to adjust relaxation parameter and convergence criterion to minimize ICC iterations. This is actually quite essential and should be accessible. The detailed debug statement for all ICC particles with old and new charges etc. is probably not worth interfacing.

**Comment by fweik** [comment]
Ok I've exposed the number of iterations needed in the last round. Is that sufficient?

**Comment by KonradBreitsprecher** [comment]
That should do it.

**Review by KonradBreitsprecher — APPROVED** [review]

**Pull request merged**